test/test_execution.py

```diff
@@ -215,39 +220,70 @@ def runner() -> int:
         )
         return execute(args)
 
-    thread = threading.Thread(target=runner)
-    thread.start()
-    retries = 20
-    pid = None
-    for i in range(retries):
-        try:
-            ps_command = "ps auxww | grep '[s]leep 60.74016230000801'"  # brackets to not match grep process
-            ps_output = subprocess.check_output(ps_command, shell=True).decode()
-            pid = int(ps_output.split()[1])
-            break
-        except subprocess.CalledProcessError as e:
-            print(f"Attempt {i} failed with msg: {e}", file=sys.stderr)
-            sleep(0.1)  # Retry after a short delay
-
-    if pid is not None:
-        os.kill(pid, signal.SIGTERM)
+    wait_time = 0.3
+    proc = multiprocessing.Process(target=runner)
+    proc.start()
+    sleep(wait_time)
+    assert proc.pid is not None, "Process PID should not be None"  # for mypy
+    os.kill(proc.pid, signal.SIGINT)
+    proc.join()
+
+    # Once the command has been killed, duct should exit gracefully with exit code 0
+    assert proc.exitcode == 0
+
+    if fail_time is None or fail_time != 0:
+        assert_expected_files(temp_output_dir, exists=False)
     else:
-        raise RuntimeError("Failed to find sleep process")
+        # proc exit code should Cannot retrieve the exit code from the thread, it is written to the file
+        with open(os.path.join(temp_output_dir, SUFFIXES["info"])) as info:
+            info_data = json.loads(info.read())
+
+        command_exit_code = info_data["execution_summary"]["exit_code"]
+        # SIGINT
+        assert command_exit_code == 128 + 2
 
-    thread.join()
+
+@pytest.mark.parametrize("fail_time", [None, 0, 10, -1, -3.14])
+def test_signal_kill(temp_output_dir: str, fail_time: float | None) -> None:
+
+    def runner() -> int:
+        script_path = str(TEST_SCRIPT_DIR / "signal_ignorer.py")
+        kws = {}
+        if fail_time is not None:
+            kws["fail_time"] = fail_time
+        args = Arguments.from_argv([script_path], output_prefix=temp_output_dir, **kws)
+        return execute(args)
+
+    wait_time = 0.8
+    proc = multiprocessing.Process(target=runner)
+    proc.start()
+    sleep(wait_time)
+    assert proc.pid is not None, "Process PID should not be None"  # for mypy
+    os.kill(proc.pid, signal.SIGINT)
+    sleep(wait_time)
```

Comment:
do we really need to wait that long too between our interruptions here or could may be proceed faster?

anything here verifies that `duct` actually tries to stop the underlying process?  may be the signal ignorer should not just sleep but do some "compute" as to switch from sleep into action, smth like

```python
t0 = time()
while time.time() - t0 < 10:
   sleep(0.01)
```
so it might response more promptly to attempts to be killed since would not just be sleeping all the time.

Reply:
Thats a good idea, maybe that will make it faster to actually catch the interrupt?

There is some race condition weirdness with the signals when using both pypy and codecov.

Reply:
One alternative to speed up the tests would be to run coverage only on one of cpython runs, then we could lower the wait times-- FWIW these passed even on pypy when cov was disabled.